plugin-scalebar/src/main/java/com/mapbox/maps/plugin/scalebar/ScaleBarImpl.kt

```diff
@@ -300,81 +326,156 @@ class ScaleBarImpl : ScaleBar, View {
+
     // When maxDistance is small (i.e. high zoom levels near the poles) then
     // the `distance` might be bigger than maxDistance. This loop will keep removing
     // bar divisions (rectCount) until it fits
     while (unitDistance * rectCount > maxDistance && rectCount > 0) {
       rectCount--
+      distance = unitDistance * rectCount
     }
     // In case the unitDistance doesn't fit at all we fallback to maxDistance (rounded to 1
     // decimal) with 1 division
     if (rectCount == 0) {
-      unitDistance = (maxDistance * 10).toInt() / 10.0F
+      unitDistance = maxDistance.toOneDecimal()
       rectCount = 1
+      distance = unitDistance
+    }
+
+    var overlapDetected = true
+    val labelTexts: MutableList<String> = mutableListOf()
+    val labelMarginsAndAnchor: MutableList<Triple<Float, Float, Float>> = mutableListOf()
+    var unitBarWidth = 0F
+    while (overlapDetected) {
+      unitDistance = distance / rectCount
+      unitBarWidth = (unitDistance / distancePerPixel).toOneDecimal()
+      labelTexts.clear()
+      labelMarginsAndAnchor.clear()
+
+      // Let's create the text label and calculate the position per each one
+      for (idx in 0..rectCount) {
+        labelTexts.add(idx, getDistanceText(unitDistance * idx, unit))
+        labelMarginsAndAnchor.add(
+          idx, calculateTextPositions(
+            labelTexts[idx],
+            unitBarWidth * idx,
+            if (idx == 0) Paint.Align.LEFT else Paint.Align.CENTER,
+            textPaint,
+            strokeWidth,
+            rightMarginPx
+          )
+        )
+      }
+
+      // Figure out if there's an overlap, by comparing each label right position with the next
+      // one left position
+      overlapDetected = false
+      for (idx in 0 until labelMarginsAndAnchor.size - 1) {
+        // `second` is the label right margin, `first` is the left margin
+        if (labelMarginsAndAnchor[idx].second >= labelMarginsAndAnchor[idx + 1].first) {
+          if (rectCount == 1) {
+            // There is a collision and we can't reduce the count, let's hide the `0` label
+            labelTexts[0] = ""
+            break
+          }
+          // There's is a collision so reduce the `rectCount` and start all over again
+          rectCount--
+          overlapDetected = true
+          break
+        }
+      }
     }
-    return Pair(unitDistance, rectCount)
+    return SegmentsConfiguration(
+      unitDistance = unitDistance,
+      unitBarWidth = unitBarWidth,
+      rectCount = rectCount,
+      labelTexts = labelTexts.toList(),
+      labelMarginsAndAnchor = labelMarginsAndAnchor.toList()
+    )
+  }
+
+  private fun calculateTextPositions(
+    text: String,
+    anchorXPx: Float,
+    alignment: Paint.Align,
+    textPaint: Paint,
+    strokeWidth: Float,
+    rightMarginPx: Int,
+  ): Triple<Float, Float, Float> {
+    var safeAnchorX = anchorXPx
+
+    val textWidthPx = textPaint.measureText(text)
```

Comment:
Wonder how fast is this nowadays, remember there were issues with the speed of text measurement, though it was probably long time ago. At the same time if it's called on every frame - maybe worth to optimize it somehow. Since the texts are actually the same most of the times - maybe introduce some cache for it?

Reply:
I could measure it but I don't think it's a problem.

I might remember wrong but I think `TextView` measurements were expensive and Android introduced [PrecomputedText](https://developer.android.com/reference/android/text/PrecomputedText) to help with it.

Reply:
I measure it in a Pixel 3 and it takes on average 0.02milliseconds (1386 times). I would not do any improvement yet.